library/pem.c

```diff
@@ -431,15 +432,20 @@ int mbedtls_pem_read_buffer(mbedtls_pem_context *ctx, const char *header, const
         }
 
         /*
-         * The result will be ASN.1 starting with a SEQUENCE tag, with 1 to 3
-         * length bytes (allow 4 to be sure) in all known use cases.
-         *
-         * Use that as a heuristic to try to detect password mismatches.
+         * The result will be ASN.1 starting with a SEQUENCE tag. Parse it
+         * with ASN.1 functions in order to:
+         * - Have an heuristic guess about password mismatches.
+         * - Update len variable to the amount of valid data inside buf.
          */
-        if (len <= 2 || buf[0] != 0x30 || buf[1] > 0x83) {
-            mbedtls_zeroize_and_free(buf, len);
-            return MBEDTLS_ERR_PEM_PASSWORD_MISMATCH;
+        unsigned char *p = buf;
+        ret = mbedtls_asn1_get_tag(&p, buf + len, &len,
+                                   MBEDTLS_ASN1_SEQUENCE | MBEDTLS_ASN1_CONSTRUCTED);
+        if (ret != 0) {
+            mbedtls_free(buf);
+            return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PEM_INVALID_DATA, ret);
         }
+        /* Add also the sequence block (tag + len) to the total amount of valid data. */
+        len += (p - buf);
```

Comment:
This ignores everything after the SEQUENCE. So it reports the total length of the payload, but it doesn't check that there is no trailing garbage.

Reply:
Correct, but if I understood correctly the "trailing garbage" that originated #8799 from #8740 was not some "real garbage" [^1], but rather the padding used by the cipher to get the proper byte size.
In my understanding a "trailing garbage" for this scenario would be either:
1. some extra data at the end of the encrypted buffer. I guess that in this case the decryption would fail because of the extra corrupted data
2. some extra data that get encrypted together with the valid SEQUENCE. However in this case, unless this extra data is ASN.1 formatted, how can I tell if it's just padding or real extra data?
The best that I can do in this case is to check that the extra data is the minimum required to get a valid block length for the cipher or longer than that. Wdyt?

[^1]: In #8740 parsing of PEM encrypted keys failed because the length reported to the RSA module was actually the length on the encrypted buffer, not the length of the unencrypted one. The difference was basically the padding added by the cipher to perform the encryption.